crates/sui-core/src/epoch/randomness.rs

```diff
@@ -0,0 +1,516 @@
+            party,
+            processed_messages: Mutex::new(BTreeMap::new()),
+            used_messages: OnceCell::new(),
+            confirmations: Mutex::new(BTreeMap::new()),
+            dkg_output: OnceCell::new(),
+        };
+        let dkg_output = tables
+            .dkg_output
+            .get(&SINGLETON_KEY)
+            .expect("typed_store should not fail");
+        if let Some(dkg_output) = dkg_output {
+            info!(
+                "random beacon: loaded existing DKG output for epoch {}",
+                committee.epoch()
+            );
+            epoch_store
+                .metrics
+                .epoch_random_beacon_dkg_num_shares
+                .set(dkg_output.shares.as_ref().map_or(0, |shares| shares.len()) as i64);
+            manager
+                .dkg_output
+                .set(dkg_output)
+                .expect("setting new OnceCell should succeed");
+        } else {
+            info!(
+                "random beacon: no existing DKG output found for epoch {}",
+                committee.epoch()
+            );
+            // Load intermediate data.
+            manager.processed_messages.lock().unwrap().extend(
+                tables
+                    .dkg_processed_messages
+                    .safe_iter()
+                    .map(|result| result.expect("typed_store should not fail")),
+            );
+            if let Some(used_messages) = tables
+                .dkg_used_messages
+                .get(&SINGLETON_KEY)
+                .expect("typed_store should not fail")
+            {
+                manager
+                    .used_messages
+                    .set(used_messages.clone())
+                    .expect("setting new OnceCell should succeed");
+            }
+            manager.confirmations.lock().unwrap().extend(
+                tables
+                    .dkg_confirmations
+                    .safe_iter()
+                    .map(|result| result.expect("typed_store should not fail")),
+            );
+        }
+
+        Some(manager)
+    }
+
+    pub fn start_dkg(&self) -> SuiResult<()> {
+        if self.used_messages.initialized() || self.dkg_output.initialized() {
+            // DKG already started (or completed).
+            return Ok(());
+        }
+
+        let msg = self.party.create_message(&mut rand::thread_rng());
+        info!(
+                "random beacon: created DKG Message with sender={}, vss_pk.degree={}, encrypted_shares.len()={}",
+                msg.sender,
+                msg.vss_pk.degree(),
+                msg.encrypted_shares.len(),
+            );
+
+        let epoch_store = self.epoch_store()?;
+        let transaction = ConsensusTransaction::new_randomness_dkg_message(epoch_store.name, &msg);
+        self.consensus_adapter
+            .submit(transaction, None, &epoch_store)?;
+
+        let mut batch = self.tables()?.dkg_processed_messages.batch();
+        self.add_message(&mut batch, msg)?;
```

Comment:
we should add it only if it's committed as one of the first 1+f messages, no? (which we don't know yet)

Reply:
That's a good point, I removed the `add_message` call, it can come through consensus